cli/cmd/uninstall.go

```diff
@@ -0,0 +1,304 @@
+	list, err := k.Apiextensions.ApiextensionsV1beta1().CustomResourceDefinitions().List(options)
+	if err != nil {
+		return "", err
+	}
+
+	if list.Items == nil {
+		return "", nil
+	}
+
+	m, err := marshalKubernetesResource(list)
+	if err != nil {
+		return "", err
+	}
+	return m, nil
+}
+
+// Although role bindings are namespaced resources in nature
+// some admin role bindings are created and persisted in the kube-system namespace and will not be deleted
+// when the namespace is deleted
+func fetchKubeSystemRoleBindings(k *k8s.KubernetesAPI, options metav1.ListOptions) (string, error) {
+	list, err := k.RbacV1().RoleBindings("kube-system").List(options)
+	if err != nil {
+		return "", err
+	}
+
+	if list.Items == nil {
+		return "", nil
+	}
+
+	m, err := marshalKubernetesResource(list)
+	if err != nil {
+		return "", err
+	}
+	return m, nil
+}
+
+func fetchNamespaceResource(k *k8s.KubernetesAPI) (string, error) {
+	obj, err := k.CoreV1().Namespaces().Get(controlPlaneNamespace, metav1.GetOptions{})
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			return "", nil
+		}
+		return "", err
+	}
+
+	m, err := marshalKubernetesResource(obj)
+	if err != nil {
+		return "", err
+	}
+	return m, nil
+}
+
+func fetchPodSecurityPolicy(k *k8s.KubernetesAPI, options metav1.ListOptions) (string, error) {
+	list, err := k.PolicyV1beta1().PodSecurityPolicies().List(options)
+	if err != nil {
+		return "", err
+	}
+
+	if list.Items == nil {
+		return "", nil
+	}
+
+	m, err := marshalKubernetesResource(list)
+	if err != nil {
+		return "", err
+	}
+	return m, nil
+}
+
+func fetchValidatingWebhooksConfiguration(k *k8s.KubernetesAPI, options metav1.ListOptions) (string, error) {
+	list, err := k.AdmissionregistrationV1beta1().ValidatingWebhookConfigurations().List(options)
+	if err != nil {
+		return "", err
+	}
+
+	if list.Items == nil {
+		return "", nil
+	}
+
+	m, err := marshalKubernetesResource(list)
+	if err != nil {
+		return "", err
+	}
+	return m, nil
+}
+
+func fetchMutatingWebhooksConfiguration(k *k8s.KubernetesAPI, options metav1.ListOptions) (string, error) {
+	list, err := k.AdmissionregistrationV1beta1().MutatingWebhookConfigurations().List(options)
+	if err != nil {
+		return "", err
+	}
+
+	if list.Items == nil {
+		return "", nil
+	}
+
+	m, err := marshalKubernetesResource(list)
+	if err != nil {
+		return "", err
+	}
+	return m, nil
+}
+func fetchAPIRegistrationResources(k *k8s.KubernetesAPI, options metav1.ListOptions) (string, error) {
+	apiClient, err := apiregistrationv1client.NewForConfig(k.Config)
+	if err != nil {
+		return "", nil
+	}
+
+	list, err := apiClient.APIServices().List(options)
+	if err != nil {
+		return "", nil
+	}
+
+	if list.Items == nil {
+		return "", nil
+	}
+
+	m, err := marshalKubernetesResource(list)
+	if err != nil {
+		return "", err
+	}
+	return m, nil
+}
+
+func marshalKubernetesResource(resource runtime.Object) (string, error) {
+	var gvk = resource.GetObjectKind().GroupVersionKind()
+	if gvk.Empty() {
```

Comment:
I'm not sure what's going on here with populating the gvk.  Can you explain this?

Reply:
@adleong of course! I realise how unnecessarily complex this looks.

Using `client-go`, whenever we fetch a resource, the `TypeMeta` seems to be stripped from it. I haven't managed to find a reason why this happens, but going through some related issues, it seems to be an intentional design; [#541 in client-go](https://github.com/kubernetes/client-go/issues/541) is an example of an issue where this problem is brought up to light.

I found a helper function that returns the gvk by using reflection on the object type, and that's the solution I went for. Prior to this, I tried to inject the gvk myself, this seemed to be a bit cleaner, however.

If there's another solution for it, I'll get it changed asap.